xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/xwiki.js

```diff
@@ -1992,3 +1992,43 @@ document.observe("xwiki:dom:loaded", function() {
   interceptXMLHttpRequest();
   interceptFetch();
 })();
+
+/**
+ * Initializes the submit-once protection on the current page.
+ * Search for all form elements with a .xwiki-submit-once class one the document (or on an updated sub-part of the dom).
+ * For each form, once the form is submitted, mark it with the .xwiki-form-submitting class and prevent new submissions.
+ *
+ *  @since 14.10.17
+ *  @since 15.5.3
+ *  @since 15.8-rc-1
+ */
+document.observe('xwiki:dom:loaded', function () {
+  /**
+   * Initializes the root element by adding event listeners to all forms with the class 'form.xwiki-submit-once'.
+   * When a form is submitted, it prevents the submission if it is already being submitted and adds a class to visually
+   * convey that the form was submitted.
+   *
+   * @param {Element} rootElement - The root element where the forms are located.
+   */
+  function init(rootElement) {
+    Array.from(rootElement.getElementsByClassName('xwiki-submit-once')).forEach(form => {
+      form.addEventListener("submit", event => {
```

Comment:
Isn't there a danger of adding the event listener twice? According to my understanding of [the documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener), anonymous functions could be added twice so I guess it would be good to either a) note on the form that the function has already been added or b) make the function non-anonymous such that it won't be added twice. Adding the function twice is dangerous from what I understand as it would prevent submitting the form.

Reply:
I'm not sure it it would have negative consequences, but it's indeed cleaner to make sure to only have one listener of this kind. Changing that.

Reply:
ok, I've improved the javascript code to prevent the handler to be registered several times on the same element.
In addition I removed any use of prototype.